### PR TITLE
Feature 83 connect to bot

### PR DESCRIPTION
**Pull request by Chi-Chu319**
Message components from metamind bot

### CODE REVIEW

**Comment by Chi-Chu319** [comment]
loading component for saving

File: src/components/generic/loading-item/loading-item.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+interface Props {
+  text: string;
+}
+
+/**
+ * Loading item component
+ */
+const Loading: React.FC<Props> = ({
+  text
+}) => {
```
Comment (lauriLukkarinen) [inline src/components/generic/loading-item/loading-item.tsx]:
```suggestion
/**
 * Loading item component
 *
 * @param props component properties
 */
const Loading: React.FC<Props> = ({ text }) => {
```

File: src/components/layouts/app-layout/app-layout.tsx [hunk 2]
```diff
@@ -123,22 +126,26 @@ class AppLayout extends React.Component<Props, State> {
...
-              { strings.header.editor }
-            </Button>
-            <Button
-              variant="text"
-              startIcon={ <PreviewIcon/> }
-              color="secondary"
-            > 
-              { strings.header.preview }
-            </Button>
+            <NavLink exact to={ `/editor/${storyData.story?.id}` } style={{ textDecoration: "none" }}>
```
Comment (lauriLukkarinen) [inline src/components/layouts/app-layout/app-layout.tsx]:
```suggestion
            <NavLink
              exact
              to={ `/editor/${storyData.story?.id}` }
              style={{ textDecoration: "none" }}
            >
```

File: src/components/layouts/app-layout/app-layout.tsx [hunk 3]
```diff
@@ -123,22 +126,26 @@ class AppLayout extends React.Component<Props, State> {
...
+              <Button
+                variant="text"
+                startIcon={ <EditorIcon/> }
+                color="secondary"
+              >
+                { strings.header.editor }
+              </Button>
+            </NavLink>
+            <NavLink exact to={ `/preview/${storyData.story?.id}` } style={{ textDecoration: "none" }}>
```
Comment (lauriLukkarinen) [inline src/components/layouts/app-layout/app-layout.tsx]:
```suggestion
            <NavLink
              exact
              to={ `/preview/${storyData.story?.id}` }
              style={{ textDecoration: "none" }}
            >
```

File: src/components/screens/editor/editor-screen.tsx [hunk 4]
```diff
@@ -686,6 +690,12 @@ const EditorScreen: React.FC<Props> = ({
       return;
     }
 
+    if (!storyLoading && storyData && storyData.story?.id === storyId){
```
Comment (lauriLukkarinen) [inline src/components/screens/editor/editor-screen.tsx]:
```suggestion
    if (!storyLoading && storyData?.story?.id === storyId) {
```

File: src/components/screens/editor/editor-screen.tsx [hunk 5]
```diff
@@ -703,6 +713,17 @@ const EditorScreen: React.FC<Props> = ({
     });
   }
 
+  /**
+   * Renders loading
+   */
+  const renderLoading = () => {
+    return (
+      <Box style={{ height: "100%", width: "100%", alignItems: "center", justifyContent: "center", display: "flex" }}>
```
Comment (lauriLukkarinen) [inline src/components/screens/editor/editor-screen.tsx]: Move inline styles to style file

File: src/components/screens/editor/editor-screen.tsx [hunk 6]
```diff
@@ -1080,13 +1101,25 @@ const EditorScreen: React.FC<Props> = ({
...
+      <AppLayout
+      keycloak={ keycloak }
+      pageTitle={ strings.loading.loading }
+    >
+      { renderLoading() }
+    </AppLayout>
+    );
+  }
+
+  const { story, knots, selectedKnot, selectedIntent, intents, trainingMaterial, scripts } = storyData;
```
Comment (lauriLukkarinen) [inline src/components/screens/editor/editor-screen.tsx]:
```suggestion
  const {
    story,
    knots,
    selectedKnot,
    selectedIntent,
    intents,
    trainingMaterial,
    scripts
  } = storyData;
```

File: src/components/screens/preview/preview-screen.styles.ts [hunk 7]
```diff
@@ -1,4 +1,16 @@
...
+
+export const usePreviewStyles = makeStyles({
+  previewContainer: {
+    // 100% height substracted with two toolbars and extra spacing for tight fit
+    height: `calc(100% - ${toolbarHeight}px - ${theme.spacing(2)}px)`,
+    width: "50%",
+    display: "flex",
+    margin: "auto"
+  }
+}, {
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.styles.ts]:
```suggestion
export const usePreviewStyles = makeStyles({

  previewContainer: {
    // 100% height substracted with two toolbars and extra spacing for tight fit
    height: `calc(100% - ${toolbarHeight}px - ${theme.spacing(2)}px)`,
    width: "50%",
    display: "flex",
    margin: "auto"
  }

}, {
```

File: src/components/screens/preview/preview-screen.tsx [hunk 8]
```diff
@@ -1,81 +1,238 @@
...
-interface State {
+  storyData?: StoryData;
+  storyLoading: boolean;
+  loadStory: () => void;
+  setStoryData: (storyData: StoryData) => void;
 }
 
 /**
  * Preview screen component
  */
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
/**
 * Preview screen component
 *
 * @param props component properties
 */
```

File: src/components/screens/preview/preview-screen.tsx [hunk 9]
```diff
@@ -1,81 +1,238 @@
...
+      message
+    ])
+  }
 
-    this.state = {}
+  /**
+   * Start the conversation 
+   */
+  const conversationStart = () => {
+    setConversationStarted((conversationStarted) => true)
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
    setConversationStarted(true);
```

File: src/components/screens/preview/preview-screen.tsx [hunk 10]
```diff
@@ -1,81 +1,238 @@
...
   }
 
   /**
-   * Component render
+   * Bot reset
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
   * Bot reset
```

File: src/components/screens/preview/preview-screen.tsx [hunk 11]
```diff
@@ -1,81 +1,238 @@
...
 
   /**
-   * Component render
+   * Bot reset 
    */
-  public render = () => {
-    const { keycloak } = this.props;
+  const botReset = () => {
+    setMessageDatas((messageDatas) => []);
+    setConversationStarted((conversationStarted) => false);
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
    setMessageDatas([]);
    setConversationStarted(false);
```

File: src/components/screens/preview/preview-screen.tsx [hunk 12]
```diff
@@ -1,81 +1,238 @@
...
-   * @param props props
+   * Bot or user response 
    */
-  constructor(props: Props) {
-    super(props);
+  const botOrUserResponse = (message: MessageData) => {
+    setMessageDatas((messageDatas) => [
+      ...messageDatas.filter(messageData => !messageData.id.startsWith("temp")), 
+      message
+    ])
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
    ]);
```

File: src/components/screens/preview/preview-screen.tsx [hunk 13]
```diff
@@ -1,81 +1,238 @@
...
+    // eslint-disable-next-line
+  }, []);
+
+  /**
+   * Interrupt the bot before the bot response 
+   */
+  const botInterrupted = () => {
+    setMessageDatas((messageDatas) =>
+      messageDatas.filter(messageData => !messageData.id.startsWith("temp"))
+    )
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
    );
```

File: src/components/screens/preview/preview-screen.tsx [hunk 14]
```diff
@@ -1,81 +1,238 @@
...
+  const botReset = () => {
+    setMessageDatas((messageDatas) => []);
+    setConversationStarted((conversationStarted) => false);
+  }
 
+  /**
+   * Update the message end 
+   */
+    const messagesEndUpdate = (messagesEnd?: HTMLDivElement) => {
+      setMessagesEnd((messageEndPrev) => messagesEnd);
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
      setMessagesEnd(messagesEnd);
```

File: src/components/screens/preview/preview-screen.tsx [hunk 15]
```diff
@@ -1,81 +1,238 @@
...
-    const { keycloak } = this.props;
+  const botReset = () => {
+    setMessageDatas((messageDatas) => []);
+    setConversationStarted((conversationStarted) => false);
+  }
 
+  /**
+   * Update the message end 
+   */
+    const messagesEndUpdate = (messagesEnd?: HTMLDivElement) => {
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
  /**
   * Update the message end
   *
   * @param messagesEnd message end
   */
    const messagesEndUpdate = (messagesEnd?: HTMLDivElement) => {
```

File: src/components/screens/preview/preview-screen.tsx [hunk 16]
```diff
@@ -1,81 +1,238 @@
...
+
+  /**
+   * Fetches knots list for the story
+   */
+  const fetchData = async () => {
+    if (!accessToken) {
+      return;
+    }
+
+    if (!storyLoading && storyData && storyData.story?.id === storyId){
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
    if (!storyLoading && storyData?.story?.id === storyId) {
```

File: src/components/screens/preview/preview-screen.tsx [hunk 17]
```diff
@@ -1,81 +1,238 @@
...
-        pageTitle="Story name"
-        dataChanged={ true }
-        storySelected={ true }
-      >
-        { this.renderLeftToolbar() }
-        <Box marginLeft="320px">
 
-        </Box>
-      </AppLayout>
+      <Box style={{ height: "100%", width: "100%", alignItems: "center", justifyContent: "center", display: "flex" }}>
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]: Move inline styles to style file

File: src/components/screens/preview/preview-screen.tsx [hunk 18]
```diff
@@ -1,81 +1,238 @@
...
+  storyData,
+  storyLoading,
+  loadStory,
+  setStoryData
+}) => {
+  const { storyId } = useParams<{ storyId: string }>();
+  const classes = usePreviewStyles();
+  const [ messageDatas, setMessageDatas ] = React.useState<MessageData[]>([]);
+  const [ conversationStarted, setConversationStarted ] = React.useState(false);
+  const [ messagesEnd, setMessagesEnd ] = React.useState<HTMLDivElement | undefined>(undefined);
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
  const [ messagesEnd, setMessagesEnd ] = React.useState<HTMLDivElement>();
```

File: src/components/screens/preview/preview-screen.tsx [hunk 19]
```diff
@@ -1,81 +1,238 @@
...
    */
-  private renderLeftToolbar = () => {
+  const renderLeftToolbar = () => {
     return (
       <Drawer
         variant="permanent"
         anchor="left"
       >
         <Toolbar/>
+        <Box height="49px" display="flex" borderBottom={1}>
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
        <Box height="49px" display="flex" borderBottom={ 1 }>
```

File: src/components/screens/preview/preview-screen.tsx [hunk 20]
```diff
@@ -1,81 +1,238 @@
...
+          <KnotPanel 
+            knots={ knots ?? [] }
+            onKnotClick={ onKnotClick }
+          />
+        </Box>
       </Drawer>
     );
   }
+
+  const renderPreviewContent = () => {
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx:176]:
```suggestion
  /**
   * Renders preview content
   */
  const renderPreviewContent = () => {
```

File: src/components/screens/preview/preview-screen.tsx [hunk 21]
```diff
@@ -1,81 +1,238 @@
...
+  /**
+   * Event handler for on knot click
+   * 
+   * @param knot knot
+   */
+  const onKnotClick = (knot: Knot) => {
+    // TODO make bot responsive
+    if (!knot?.coordinates?.x || !knot?.coordinates?.y) {
+      return;
+    }
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx:214]:
```suggestion
    }

```

File: src/components/screens/preview/preview-screen.tsx [hunk 22]
```diff
@@ -1,81 +1,238 @@
...
+      <AppLayout
+      keycloak={ keycloak }
+      pageTitle={ strings.loading.loading }
+    >
+      { renderLoading() }
+    </AppLayout>
+    );
+  }
+
+  const { story, knots, selectedKnot, selectedIntent, intents, trainingMaterial } = storyData;
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
  const {
    story,
    knots,
    selectedKnot,
    selectedIntent,
    intents,
    trainingMaterial
  } = storyData;
```

File: src/components/views/story-preview-view/story-preview-view.tsx [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+   */
+  const restartConversation = () => {
+    // TODO 
+  }
+
+  /**
+   * Load the story
+   */
+  const loadData = () => {
+    if (messageDatas.length !== 0) {
```
Comment (lauriLukkarinen) [inline src/components/views/story-preview-view/story-preview-view.tsx]:
```suggestion
    if (messageDatas.length) {
```

File: src/localization/strings.ts [hunk 24]
```diff
@@ -125,6 +125,35 @@ export interface IStrings extends LocalizedStringsMethods {
     }
   };
 
+  previewScreen: {
```
Comment (lauriLukkarinen) [inline src/localization/strings.ts:131]:
```suggestion
  /**
   * Translations related to preview screen
   */
  previewScreen: {
```

File: src/reducers/story.ts [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+import { StoryData } from "../types";
+
+/**
+ * Redux story state
+ */
+export interface StoryState {
+  storyLoading: boolean;
+  storyData?: StoryData;
+}
+
```
Comment (lauriLukkarinen) [inline src/reducers/story.ts]:
```suggestion
```

File: src/reducers/story.ts [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+        storyData: undefined 
+      };
+    case SET_STORY:
+      const storyData = action.storyData;
+
+      return { 
+        ...state,
+        storyData,
+        storyLoading: false 
+      };
```
Comment (lauriLukkarinen) [inline src/reducers/story.ts]:
```suggestion
      return { 
        ...state,
        storyData: action.storyData,
        storyLoading: false 
      };
```

**Review by lauriLukkarinen — CHANGES_REQUESTED** [review]

File: src/components/screens/editor/editor-screen.tsx [hunk 27]
```diff
@@ -1080,13 +1101,33 @@ const EditorScreen: React.FC<Props> = ({
...
   }
 
+  if (storyLoading || !storyData) {
+    return (
+      <AppLayout
+      keycloak={ keycloak }
+      pageTitle={ strings.loading.loading }
+    >
+      { renderLoading() }
+    </AppLayout>
```
Comment (lauriLukkarinen) [inline src/components/screens/editor/editor-screen.tsx]:
```suggestion
      <AppLayout
        keycloak={ keycloak }
        pageTitle={ strings.loading.loading }
      >
        { renderLoading() }
      </AppLayout>
```

File: src/components/screens/preview/preview-screen.tsx [hunk 28]
```diff
@@ -1,81 +1,252 @@
...
-import { History } from "history";
+import { usePreviewStyles } from "./preview-screen.styles";
 import { KeycloakInstance } from 'keycloak-js';
+import KnotPanel from "../../knot-components/knot-list/knot-list";
+import { Story, Knot, Intent, TrainingMaterial } from "../../../generated/client/models";
+import strings from "../../../localization/strings";
+import { Typography } from "@material-ui/core";
+import Api from "../../../api/api";
+import { useParams } from "react-router-dom";
+import { StoryData } from "../../../types"
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
import { StoryData } from "../../../types";
```

File: src/components/screens/preview/preview-screen.tsx [hunk 29]
```diff
@@ -1,81 +1,252 @@
...
+  React.useEffect(() => {
+    fetchData();
+    // eslint-disable-next-line
+  }, []);
+
+  /**
+   * Interrupt the bot before the bot response 
+   */
+  const botInterrupted = () => {
+    setMessageDatas((messageDatas) =>
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
    setMessageDatas(messageDatas =>
```

File: src/components/screens/preview/preview-screen.tsx [hunk 30]
```diff
@@ -1,81 +1,252 @@
...
+      messageDatas.filter(messageData => !messageData.id.startsWith("temp"))
+    );
+  }
+
+  /**
+   * Bot or user response 
+   */
+  const botOrUserResponse = (message: MessageData) => {
+    setMessageDatas((messageDatas) => [
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
    setMessageDatas(messageDatas => [
```

File: src/components/screens/preview/preview-screen.tsx [hunk 31]
```diff
@@ -1,81 +1,252 @@
...
+      <AppLayout
+      keycloak={ keycloak }
+      pageTitle={ strings.loading.loading }
+    >
+      { renderLoading() }
+    </AppLayout>
+    );
+  }
+
+  const { story,
```
Comment (lauriLukkarinen) [inline src/components/screens/preview/preview-screen.tsx]:
```suggestion
  const {
    story,
```

File: src/components/views/story-preview-view/story-preview-view.tsx [hunk 32]
```diff
@@ -0,0 +1,131 @@
+import * as React from "react";
+import { useStoryPreviewViewStyles } from "./story-preview-view";
+import { MessageData } from "metamind-metatavu-bot/dist/types";
+import MessageList from "metamind-metatavu-bot/dist/components/message-list/message-list"
+import MessageInput from "metamind-metatavu-bot/dist/components/message-input/message-input"
```
Comment (lauriLukkarinen) [inline src/components/views/story-preview-view/story-preview-view.tsx]:
```suggestion
import MessageList from "metamind-metatavu-bot/dist/components/message-list/message-list";
import MessageInput from "metamind-metatavu-bot/dist/components/message-input/message-input";
```

**Review by lauriLukkarinen — APPROVED** [review]

**Pull request merged**